h/auth/util.py

```diff
@@ -157,25 +149,42 @@ def check_auth_client(username, password, request):
     if not hmac.compare_digest(client.secret, client_secret):
         return None
 
-    return principals_for_auth_client(client)
+    return client
 
 
 def principals_for_auth_client(client):
     """
     Return the list of additional principals for an auth client
 
-    :type client: h.models.AuthClient
+    :type client: :py:class:`h.models.auth_client.AuthClient`
     :rtype: list
     """
 
     principals = set([])
 
-    principals.add('auth_client:{authority}'.format(authority=client.authority))
+    principals.add('client:{client_id}@{authority}'.format(client_id=client.id, authority=client.authority))
     principals.add('authority:{authority}'.format(authority=client.authority))
 
     return list(principals)
 
 
+def principals_for_auth_client_user(user, client):
+    """
+    Return a union of client and user principals for forwarded user
+
+    :type user: :py:class:`h.models.user.User`
+    :type client: :py:class:`h.models.auth_client.AuthClient`
+    :rtype: list
+    """
+    user_principals = principals_for_user(user)
+    client_principals = principals_for_auth_client(client)
+
+    all_principals = user_principals + client_principals
+    distinct_principals = list(set(all_principals))
+
+    return distinct_principals
```

Comment:
I'm wondering why an `X-Forwarded-User` needs the principals for both the user and the auth client, instead of just the principals for the user?

Reply:
I want to be able to "mark" a request as having been authenticated via an auth_client. Otherwise there is no easy way to discern a forwarded-user auth'd request from a Token auth'd request.